C++/Navio/Common/gpio.cpp

```diff
@@ -137,37 +135,47 @@ void Pin::toggle()
     write(!read());
 }
 
+uint32_t Pin::getPeriBase() const
+{
+    int rpi_version = getRaspberryPiVersion();
+
+    switch (rpi_version) {
+      case 1:
+        return BCM2708_PERI_BASE;
+      case 2:
+        return BCM2709_PERI_BASE;
+      case 3:
+        return BCM2835_PERI_BASE;
+      default:
+        std::cerr << "Unexpected version. Setting up configurations as on RPI2/3" << std::endl;
+        return BCM2709_PERI_BASE;
+    }
+}
+
+
+
 int Pin::getRaspberryPiVersion() const
 {
-    char buffer[MAX_SIZE_LINE];
-    const char* hardware_description_entry = "Hardware";
-    const char* v1 = "BCM2708";
-    const char* v2 = "BCM2709";
-    const char* v3 = "BCM2835";
-    char* flag;
-    FILE* fd;
-
-    fd = fopen("/proc/cpuinfo", "r");
-
-    while (fgets(buffer, MAX_SIZE_LINE, fd) != NULL) {
-        flag = strstr(buffer, hardware_description_entry);
-
-        if (flag != NULL) {
-            if (strstr(buffer, v3) != NULL) {
-                fclose(fd);
-                return 3;
-            } else if (strstr(buffer, v2) != NULL) {
-                fclose(fd);
-                return 2;
-            } else if (strstr(buffer, v1) != NULL) {
-                fclose(fd);
-                return 1;
-            }
-        }
+    int rpi_version;
+
+    std::string buffer;
+    std::ifstream f(SYSFS_RPI_MODEL_FILE_PATH);
+
+    if (!f) {
+      std::cerr << "Could not open " << SYSFS_RPI_MODEL_FILE_PATH << std::endl;
+      return ERROR;
     }
 
-    /* defaults to 1 */
-    fprintf(stderr, "Could not detect RPi version, defaulting to 1\n");
-    fclose(fd);
-    return 1;
+    while (f >> buffer) {
+      try {
+        rpi_version = stoi(buffer);
+        return rpi_version;
+      }
+      catch(std::invalid_argument) {}
+    }
+
+    f.close();
```

Comment:
Oh, and we can omit this. RAII deals with it better.

Reply:
Sure, thank you!